include/private/net/libssl.hpp

```diff
@@ -0,0 +1,331 @@
+
+        To avoid the cache potentially becoming too large (perhaps also because
+        of bugs) we have set a large hard limit on the maximum number of
+        `SSL_CTX *` that we keep open concurrently.
+
+        2. since assuming OpenSSL will correctly behave in a MT environment
+        seem to be an over-assumption (see above for links), I decided to stay
+        on the safe side an make the cache thread local.
+
+        After #1297 is merged, this means that every "big" task (i.e. a test,
+        orchestration, etc) will use its own thread local cache.
+
+        Overall, this choice is a bit more defensive than needed, perhaps, but
+        I do also see some advantages in not keeping `SSL_CTX *` - and hence
+        all related caching - alive "forever". So I think it's okay.
+    */
+
+    /// Return thread local instance of the cache.
+    static Cache &thread_local_instance() {
+        static thread_local Cache instance;
+        return instance;
+    }
+
+    /*!
+        \brief Factory to construct an `SSL *`.
+
+        \param ca_bundle_path The CA bundle path to use. This should be set
+        using the `net/ca_bundle_path` configuration variable.
+
+        \param hostname Hostname for which you want to do SSL and which will
+        be used in the SNI extension.
+
+        \param logger Logger used to log messages.
+
+        \return Error on failure (the returned object will be false when
+        evaluated in a boolean context) or `SSL *` on success.
+
+        \remark This factory gives you the ownership of the returned
+        `SSL *` object and it is your responsibility to dispose of it
+        when done. In practice, when using libevent, you will typically
+        pass the `SSL *` to a bufferevent that will own it.
+
+        \remark This operation is thread safe by definition since the
+        cache is created using thread-local storage.
+    */
+    ErrorOr<SSL *> get_client_ssl(std::string ca_bundle_path,
+                                  std::string hostname, Var<Logger> logger) {
+        /*
+           Implementation note: `SSL_CTX *` are reference counted and they will
+           be alive as long as there is an `SSL *` using them. As such, we do
+           not need to worry about `all_.clear()` causing `SSL_CTX_free` to be
+           called on the `SSL_CTX *` we indirectly manage, because the usage
+           count semantic will keep us safe from premature frees.
+
+           See above for a specific link re: refcounting.
+        */
+        if (all_.size() >= max_cache_size) {
+            logger->warn("ssl: hit hard limit of maximum cached SSL_CTX");
+            all_.clear();
+        }
+        if (all_.count(ca_bundle_path) == 0) {
+            ErrorOr<Var<Context>> maybe_context =
+                  Context::make(ca_bundle_path, logger);
+            if (!maybe_context) {
+                return maybe_context.as_error();
```

Comment:
This case is not covered by the unit tests. To cover it, we need to wrap `Context::make` into an inline function and make the current method a template using that function or another function with same signature. That seems easy.